File: sdk-php/tests/MetricsTest.php
<?php

declare(strict_types=1);

namespace Flux\Tests;

use Flux\ErrorClass;
use Flux\Metrics\ConnectionState;
use Flux\Metrics\ConsumeOutcome;
use Flux\Metrics\InMemoryMetrics;
use Flux\Metrics\MetricsSink;
use Flux\Metrics\NoMetrics;
use Flux\Metrics\PublishOutcome;
use Flux\Protocol;
use PHPUnit\Framework\TestCase;

/**
 * Métricas — specification/08-observability.md.
 *
 * Los nombres y las etiquetas son **contrato entre SDKs**, así que se comprueban sobre la
 * salida real: si el de PHP nombrara distinto la tasa de DLQ, la de los servicios PHP no
 * se podría sumar con la de los de Go y un panel del ecosistema sería imposible.
 */
final class MetricsTest extends TestCase
{
    private const SUBJECT = 'pedidos.pedido.v1.creado';
    private const CONSUMER = 'facturacion-api__pedidos_pedido_v1_creado';

    // ─── Buckets — §3 ─────────────────────────────────────────────────────────

    /** La invariante que da sentido al histograma entero. */
    public function testElUltimoBucketEsElAckWait(): void
    {
        // Un handler en el bucket superior está a punto de que su mensaje se reentregue
        // mientras aún se ejecuta, así que `…_bucket{le="30"}` frente al total mide
        // directamente cuántos eventos rozan la ejecución concurrente. Si el bucket dejara
        // de coincidir con el plazo real, mediría algo que no le importa a nadie — y esa
        // desincronización solo se detecta aquí.
        $buckets = MetricsSink::DURATION_BUCKETS;

        self::assertSame(
            (float) Protocol::ACK_WAIT_MS / 1000.0,
            (float) $buckets[count($buckets) - 1],
        );
    }

    public function testLosBucketsSonLosDelContrato(): void
    {
        // `protocol.json` se lee del repositorio: una divergencia entre el SDK y el
        // contrato falla en CI, no en producción con otro SDK del ecosistema.
        $protocol = ProtocolFixture::protocol();

        self::assertSame(
            array_map(floatval(...), $protocol['observability']['durationBucketsSeconds']),
            array_map(floatval(...), MetricsSink::DURATION_BUCKETS),
        );
    }

    public function testLosBucketsAscienden(): void
    {
        $buckets = MetricsSink::DURATION_BUCKETS;

        for ($i = 1; $i < count($buckets); $i++) {
            self::assertGreaterThan($buckets[$i - 1], $buckets[$i]);
        }
    }

    // ─── Nombres y etiquetas — §2 ─────────────────────────────────────────────

    public function testLasSieteMetricasSonLasDelContrato(): void
    {
        $m = new InMemoryMetrics();
        $m->eventPublished(self::SUBJECT, PublishOutcome::Ok);
        $m->eventConsumed(self::SUBJECT, self::CONSUMER, ConsumeOutcome::Ok);
        $m->handlerDuration(self::SUBJECT, self::CONSUMER, 0.4);
        $m->eventDlq(self::SUBJECT, self::CONSUMER, ErrorClass::Permanent, 'HTTP_404');
        $m->eventRetried(self::SUBJECT, self::CONSUMER, 3);
        $m->consumerPending(self::SUBJECT, self::CONSUMER, 42);
        $m->connectionState(ConnectionState::Connected);

        $salida = $m->render();
        $declaradas = array_keys(ProtocolFixture::protocol()['observability']['metrics']);
        self::assertCount(7, $declaradas);

        foreach ($declaradas as $nombre) {
            self::assertStringContainsString($nombre, $salida, "falta {$nombre}");
        }
    }

    public function testCuentaPublicacionesPorSubjectYResultado(): void
    {
        $m = new InMemoryMetrics();
        $m->eventPublished(self::SUBJECT, PublishOutcome::Ok);
        $m->eventPublished(self::SUBJECT, PublishOutcome::Ok);
        $m->eventPublished(self::SUBJECT, PublishOutcome::Error);

        self::assertSame(
            2,
            $m->counter('flux_events_published_total{subject="' . self::SUBJECT . '",outcome="ok"}'),
        );
        self::assertSame(
            1,
            $m->counter('flux_events_published_total{subject="' . self::SUBJECT . '",outcome="error"}'),
        );
    }

    public function testLaClaveDeUnaSerieEsEstableEntreRecolectores(): void
    {
        // Sin orden estable, la misma serie temporal aparecería con dos claves según el
        // orden en que se construyeron las etiquetas, y los contadores se partirían en dos.
        $a = new InMemoryMetrics();
        $b = new InMemoryMetrics();
        $a->eventDlq(self::SUBJECT, self::CONSUMER, ErrorClass::Permanent, 'HTTP_404');
        $b->eventDlq(self::SUBJECT, self::CONSUMER, ErrorClass::Permanent, 'HTTP_404');

        self::assertSame($a->render(), $b->render());
    }

    /**
     * §2.2: **NUNCA** se etiqueta con `tenantid`, `id` ni `correlationid`.
     *
     * Aquí no puede pasar porque no hay dónde meterlos —las firmas de `MetricsSink` no lo
     * permiten—, pero el test lo fija sobre la salida por si alguien añade una etiqueta de
     * más. La cardinalidad no avisa: funciona con tres tenants y muere con diez mil.
     */
    public function testNingunaEtiquetaProhibidaLlegaALaSalida(): void
    {
        $m = new InMemoryMetrics();
        $m->eventPublished(self::SUBJECT, PublishOutcome::Ok);
        $m->eventConsumed(self::SUBJECT, self::CONSUMER, ConsumeOutcome::Ok);
        $m->eventDlq(self::SUBJECT, self::CONSUMER, ErrorClass::Poison, 'MALFORMED_JSON');
        $m->eventRetried(self::SUBJECT, self::CONSUMER, 1);
        $m->consumerPending(self::SUBJECT, self::CONSUMER, 7);
        $m->handlerDuration(self::SUBJECT, self::CONSUMER, 1.0);
        $m->connectionState(ConnectionState::Reconnecting);

        $salida = $m->render();

        foreach (ProtocolFixture::protocol()['observability']['forbiddenLabels'] as $prohibida) {
            self::assertStringNotContainsString("{$prohibida}=", $salida);
        }
    }

    // ─── Formato de exposición ────────────────────────────────────────────────

    public function testElHistogramaAcumulaEnTodosLosBucketsQueSuperanElValor(): void
    {
        $m = new InMemoryMetrics();
        $m->handlerDuration(self::SUBJECT, self::CONSUMER, 0.03); // por encima de 0.025

        $salida = $m->render();
        self::assertStringContainsString('le="0.025"} 0', $salida);
        self::assertStringContainsString('le="0.05"} 1', $salida);
        self::assertStringContainsString('le="+Inf"} 1', $salida);
        self::assertStringContainsString('_count{', $salida);
    }

    /**
     * `le="30"`, no `le="30.0"`: para Prometheus son **dos series distintas** y el
     * dashboard del ecosistema agrupa por la primera.
     */
    public function testElBucketEnteroSeImprimeSinDecimales(): void
    {
        $m = new InMemoryMetrics();
        $m->handlerDuration(self::SUBJECT, self::CONSUMER, 0.4);

        $salida = $m->render();
        self::assertStringContainsString('le="30"}', $salida);
        self::assertStringNotContainsString('le="30.0"}', $salida);
        self::assertStringContainsString('le="0.005"}', $salida);
    }

    public function testUnGaugeSinEtiquetasNoDejaLlavesVacias(): void
    {
        $m = new InMemoryMetrics();
        $m->connectionState(ConnectionState::Connected);

        self::assertContains('flux_connection_state 1', explode("\n", $m->render()));
    }

    public function testLosTresEstadosDeConexionSonLosDeLaSpec(): void
    {
        $declarados = ProtocolFixture::protocol()['observability']['metrics']
            ['flux_connection_state']['values'];

        self::assertSame('desconectado', $declarados[(string) ConnectionState::Disconnected->value]);
        self::assertSame('conectado', $declarados[(string) ConnectionState::Connected->value]);
        self::assertSame('reconectando', $declarados[(string) ConnectionState::Reconnecting->value]);
    }

    /** Cada línea del formato de exposición, una a una. */
    public function testLasLineasTienenFormaValidaDePrometheus(): void
    {
        $m = new InMemoryMetrics();
        $m->eventPublished(self::SUBJECT, PublishOutcome::Ok);
        $m->eventConsumed(self::SUBJECT, self::CONSUMER, ConsumeOutcome::Ok);
        $m->eventDlq(self::SUBJECT, self::CONSUMER, ErrorClass::Permanent, 'HTTP_404');
        $m->eventRetried(self::SUBJECT, self::CONSUMER, 3);
        $m->consumerPending(self::SUBJECT, self::CONSUMER, 42);
        $m->connectionState(ConnectionState::Connected);
        $m->handlerDuration(self::SUBJECT, self::CONSUMER, 0.4);

        foreach (explode("\n", $m->render()) as $linea) {
            if ($linea === '' || str_starts_with($linea, '#')) {
                continue;
            }

            self::assertMatchesRegularExpression(
                '/^flux_[a-z0-9_]+(\{[^}]*\})? -?[0-9.]+$/',
                $linea,
                "línea no válida para Prometheus: {$linea}"
            );
        }
    }

    /**
     * Un `code` con comillas rompería el formato y **Prometheus descartaría el scrape
     * entero**, no solo esa línea. Los `code` salen de errores de terceros, así que la
     * comilla llega antes o después.
     */
    public function testEscapaLasComillasDeLosValoresDeEtiqueta(): void
    {
        $m = new InMemoryMetrics();
        $m->eventDlq(self::SUBJECT, self::CONSUMER, ErrorClass::Permanent, 'con "comillas"');

        $linea = null;
        foreach (explode("\n", $m->render()) as $l) {
            if (str_starts_with($l, 'flux_events_dlq_total')) {
                $linea = $l;
            }
        }

        self::assertNotNull($linea);
        self::assertStringContainsString('code="con \\"comillas\\""', $linea);
        // Las comillas SIN escapar quedan balanceadas: 2 por cada una de las 4 etiquetas.
        self::assertSame(0, preg_match_all('/(?<!\\\\)"/', $linea) % 2, $linea);
    }

    public function testEscapaBarrasYSaltosDeLinea(): void
    {
        $m = new InMemoryMetrics();
        $m->eventDlq(self::SUBJECT, self::CONSUMER, ErrorClass::Poison, "a\\b\nc");

        self::assertStringContainsString('code="a\\\\b\\nc"', $m->render());
    }

    public function testElIntentoSeEtiquetaComoTexto(): void
    {
        $m = new InMemoryMetrics();
        $m->eventRetried(self::SUBJECT, self::CONSUMER, 3);

        self::assertSame(1, $m->counter(
            'flux_events_retried_total{subject="' . self::SUBJECT . '",consumer="'
            . self::CONSUMER . '",attempt="3"}'
        ));
    }

    public function testElGaugeDePendientesSeSobrescribeNoSeAcumula(): void
    {
        $m = new InMemoryMetrics();
        $m->consumerPending(self::SUBJECT, self::CONSUMER, 10);
        $m->consumerPending(self::SUBJECT, self::CONSUMER, 3);

        self::assertSame(3.0, $m->gauge(
            'flux_consumer_pending{subject="' . self::SUBJECT . '",consumer="' . self::CONSUMER . '"}'
        ));
    }

    // ─── Literales de las etiquetas — §2.1 ────────────────────────────────────

    public function testLosLiteralesDeLasEtiquetasSonLosDeLaSpec(): void
    {
        self::assertSame(
            ['ok', 'retryable', 'permanent', 'poison', 'invalid_schema', 'invalid_signature'],
            array_map(static fn (ConsumeOutcome $o): string => $o->value, ConsumeOutcome::cases()),
        );
        self::assertSame(
            ['ok', 'invalid_schema', 'error'],
            array_map(static fn (PublishOutcome $o): string => $o->value, PublishOutcome::cases()),
        );
        // `reason` reutiliza ErrorClass: su valor ES el literal de `dlqreason`, así que no
        // hay un segundo enum que alguien pueda desincronizar.
        self::assertSame(
            ['retryable', 'permanent', 'poison'],
            array_map(static fn (ErrorClass $c): string => $c->value, ErrorClass::cases()),
        );
    }

    public function testLaClaseDeErrorSeTraduceAlOutcomeCorrespondiente(): void
    {
        self::assertSame(
            ConsumeOutcome::Retryable,
            ConsumeOutcome::fromErrorClass(ErrorClass::Retryable),
        );
        self::assertSame(
            ConsumeOutcome::Permanent,
            ConsumeOutcome::fromErrorClass(ErrorClass::Permanent),
        );
        self::assertSame(
            ConsumeOutcome::Poison,
            ConsumeOutcome::fromErrorClass(ErrorClass::Poison),
        );
    }

    // ─── El default ───────────────────────────────────────────────────────────

    public function testNoMetricsNoGuardaNadaYNoLanza(): void
    {
        // Un SDK no debe imponer un backend de métricas: el default no hace nada.
        $m = new NoMetrics();
        $m->eventPublished(self::SUBJECT, PublishOutcome::Ok);
        $m->eventConsumed(self::SUBJECT, self::CONSUMER, ConsumeOutcome::Poison);
        $m->handlerDuration(self::SUBJECT, self::CONSUMER, 0.1);
        $m->eventDlq(self::SUBJECT, self::CONSUMER, ErrorClass::Poison, 'X');
        $m->eventRetried(self::SUBJECT, self::CONSUMER, 1);
        $m->consumerPending(self::SUBJECT, self::CONSUMER, 0);
        $m->connectionState(ConnectionState::Disconnected);

        $this->addToAssertionCount(1);
    }
}

File: sdk-go/metrics_test.go
package flux

import (
	"encoding/json"
	"os"
	"regexp"
	"sort"
	"strings"
	"sync"
	"testing"
)

// Réplica de sdk-node/test/metrics.test.ts más los casos que el contrato pide y que solo
// un test puede sostener: que los nombres y las etiquetas son los de protocol.json, y que
// ninguna etiqueta es `tenantid`.

// lineaPrometheus es la gramática del formato de exposición: `nombre{etiqueta="valor",…}
// valor`. Las comillas dentro de un valor solo valen escapadas — de ahí `(?:[^"\\]|\\.)*`.
var lineaPrometheus = regexp.MustCompile(
	`^[a-zA-Z_:][a-zA-Z0-9_:]*` +
		`(\{[a-zA-Z_][a-zA-Z0-9_]*="(?:[^"\\]|\\.)*"` +
		`(,[a-zA-Z_][a-zA-Z0-9_]*="(?:[^"\\]|\\.)*")*\})?` +
		` -?[0-9.]+$`)

var serieRe = regexp.MustCompile(`^([a-zA-Z_:][a-zA-Z0-9_:]*)(?:\{(.*)\})? `)
var etiquetaRe = regexp.MustCompile(`([a-zA-Z_][a-zA-Z0-9_]*)="`)

const (
	subjectPrueba  = "pedidos.pedido.v1.creado"
	consumerPrueba = "facturacion-api__pedidos_pedido_v1_creado"
)

// protocolJSON lee el contrato desde la raíz del repo.
//
// El SDK y protocol.json pueden divergir en un rename descuidado, y el síntoma sería un
// panel vacío, no un error. Mejor que falle aquí.
func protocolJSON(t *testing.T) map[string]any {
	t.Helper()
	raw, err := os.ReadFile("../protocol.json")
	if err != nil {
		t.Fatalf("no se pudo leer protocol.json: %v", err)
	}
	var doc map[string]any
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatalf("protocol.json no es JSON válido: %v", err)
	}
	return doc
}

func observabilidad(t *testing.T) map[string]any {
	t.Helper()
	return protocolJSON(t)["observability"].(map[string]any)
}

// conTodasLasMetricas registra una muestra de cada una de las siete.
func conTodasLasMetricas() *InMemoryMetrics {
	m := NewInMemoryMetrics()
	m.EventPublished(subjectPrueba, PublishOK)
	m.EventConsumed(subjectPrueba, consumerPrueba, ConsumeOK)
	m.HandlerDuration(subjectPrueba, consumerPrueba, 0.4)
	m.EventDLQ(subjectPrueba, consumerPrueba, DLQReasonPermanent, "HTTP_404")
	m.EventRetried(subjectPrueba, consumerPrueba, 3)
	m.ConsumerPending(subjectPrueba, consumerPrueba, 42)
	m.ConnectionState(StateConnected)
	return m
}

// series convierte el texto expuesto en {nombre: {etiquetas}}.
func series(t *testing.T, render string) map[string]map[string]bool {
	t.Helper()
	out := map[string]map[string]bool{}
	for _, linea := range strings.Split(strings.TrimRight(render, "\n"), "\n") {
		if linea == "" || strings.HasPrefix(linea, "#") {
			continue
		}
		m := serieRe.FindStringSubmatch(linea)
		if m == nil {
			t.Fatalf("línea no parseable: %q", linea)
		}
		if out[m[1]] == nil {
			out[m[1]] = map[string]bool{}
		}
		for _, e := range etiquetaRe.FindAllStringSubmatch(m[2], -1) {
			out[m[1]][e[1]] = true
		}
	}
	return out
}

func claves(m map[string]bool) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

// ─── El contrato ─────────────────────────────────────────────────────────────

func TestNombresDeMetricasSonLosDeProtocolJSON(t *testing.T) {
	// Si dos SDKs nombran distinto lo mismo, agrupar deja de funcionar en cuanto el
	// ecosistema es polyglot — que es siempre (08-observability.md §1).
	expuestas := map[string]bool{}
	for nombre := range series(t, conTodasLasMetricas().Render()) {
		// El histograma se expone en tres familias derivadas del mismo nombre.
		nombre = strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(
			nombre, "_bucket"), "_sum"), "_count")
		expuestas[nombre] = true
	}

	esperadas := map[string]bool{}
	for nombre := range observabilidad(t)["metrics"].(map[string]any) {
		esperadas[nombre] = true
	}

	if strings.Join(claves(expuestas), ",") != strings.Join(claves(esperadas), ",") {
		t.Errorf("métricas expuestas %v, protocol.json declara %v",
			claves(expuestas), claves(esperadas))
	}
}

func TestEtiquetasSonLasDeProtocolJSON(t *testing.T) {
	expuestas := series(t, conTodasLasMetricas().Render())
	for nombre, def := range observabilidad(t)["metrics"].(map[string]any) {
		d := def.(map[string]any)
		var esperadas []string
		for _, l := range d["labels"].([]any) {
			esperadas = append(esperadas, l.(string))
		}
		sort.Strings(esperadas)

		clave := nombre
		if d["type"] == "histogram" {
			clave = nombre + "_count" // _bucket añade `le`, que es del formato, no una etiqueta
		}
		got := claves(expuestas[clave])
		if strings.Join(got, ",") != strings.Join(esperadas, ",") {
			t.Errorf("%s expone %v, protocol.json declara %v", nombre, got, esperadas)
		}
	}
}

func TestNingunaEtiquetaEsDeAltaCardinalidad(t *testing.T) {
	// NUNCA se etiqueta por tenantid, id ni correlationid: un tenant nuevo no debe crear
	// series temporales nuevas. Para eso están las trazas, donde la cardinalidad de un
	// evento individual no es un problema — 08-observability.md §2.2.
	prohibidas := map[string]bool{}
	for _, l := range observabilidad(t)["forbiddenLabels"].([]any) {
		prohibidas[l.(string)] = true
	}
	for nombre, etiquetas := range series(t, conTodasLasMetricas().Render()) {
		for etiqueta := range etiquetas {
			if prohibidas[etiqueta] {
				t.Errorf("%s está etiquetada por %q", nombre, etiqueta)
			}
		}
	}
}

// ─── Buckets ─────────────────────────────────────────────────────────────────

func TestUltimoBucketEsElAckWait(t *testing.T) {
	// 08-observability.md §3: un handler en el bucket superior está a punto de que su
	// mensaje se reentregue mientras aún se ejecuta. Si el bucket deja de coincidir con
	// el plazo real, mide algo que no le importa a nadie — y el día que alguien cambie
	// DefaultAckWait nadie se acordará de este número si no falla un test.
	buckets := DurationBuckets()
	if got := buckets[len(buckets)-1]; got != DefaultAckWait.Seconds() {
		t.Errorf("último bucket = %v s, ack_wait = %v s", got, DefaultAckWait.Seconds())
	}
}

func TestBucketsCoincidenConProtocolJSON(t *testing.T) {
	esperados := observabilidad(t)["durationBucketsSeconds"].([]any)
	buckets := DurationBuckets()
	if len(buckets) != len(esperados) {
		t.Fatalf("%d buckets, protocol.json declara %d", len(buckets), len(esperados))
	}
	for i, want := range esperados {
		if buckets[i] != want.(float64) {
			t.Errorf("bucket[%d] = %v, se esperaba %v", i, buckets[i], want)
		}
	}
}

func TestBucketsAscendentes(t *testing.T) {
	buckets := DurationBuckets()
	for i := 1; i < len(buckets); i++ {
		if buckets[i] <= buckets[i-1] {
			t.Errorf("bucket[%d]=%v no es mayor que bucket[%d]=%v",
				i, buckets[i], i-1, buckets[i-1])
		}
	}
}

func TestDurationBucketsDevuelveUnaCopia(t *testing.T) {
	// Un slice a nivel de paquete sería mutable desde fuera y una entrada alterada
	// cambiaría en silencio el histograma de todo consumidor creado después.
	DurationBuckets()[0] = 999
	if DurationBuckets()[0] != 0.005 {
		t.Error("DurationBuckets comparte el array subyacente entre llamadas")
	}
}

// ─── Recolector ──────────────────────────────────────────────────────────────

func TestCuentaPublicacionesPorSubjectYResultado(t *testing.T) {
	m := NewInMemoryMetrics()
	m.EventPublished(subjectPrueba, PublishOK)
	m.EventPublished(subjectPrueba, PublishOK)
	m.EventPublished(subjectPrueba, PublishInvalidSchema)

	counters, _ := m.Snapshot()
	if got := counters[`flux_events_published_total{outcome="ok",subject="`+subjectPrueba+`"}`]; got != 2 {
		t.Errorf("ok = %v, se esperaba 2", got)
	}
	if got := counters[`flux_events_published_total{outcome="invalid_schema",subject="`+subjectPrueba+`"}`]; got != 1 {
		t.Errorf("invalid_schema = %v, se esperaba 1", got)
	}
}

func TestEtiquetasOrdenadasParaQueLaClaveSeaEstable(t *testing.T) {
	// Sin orden estable, la misma serie temporal aparecería con dos claves según el orden
	// en que se construyó la lista de etiquetas.
	a, b := NewInMemoryMetrics(), NewInMemoryMetrics()
	a.EventDLQ("s", "c", DLQReasonPermanent, "X")
	b.EventDLQ("s", "c", DLQReasonPermanent, "X")
	ca, _ := a.Snapshot()
	cb, _ := b.Snapshot()
	for k := range ca {
		if _, ok := cb[k]; !ok {
			t.Errorf("clave inestable: %q", k)
		}
	}
}

func TestHistogramaAcumulaEnTodosLosBucketsQueSuperanElValor(t *testing.T) {
	m := NewInMemoryMetrics()
	m.HandlerDuration("s", "c", 0.03) // cae por encima de 0.025
	salida := m.Render()

	for _, want := range []string{`le="0.025"} 0`, `le="0.05"} 1`, `le="+Inf"} 1`} {
		if !strings.Contains(salida, want) {
			t.Errorf("falta %q en:\n%s", want, salida)
		}
	}
}

func TestHandlerLentoCaeEnElBucketDelAckWait(t *testing.T) {
	// La señal que el bucket superior existe para dar: 29 s cuenta en le="30" y nada
	// antes; el evento está a un segundo de ejecutarse consigo mismo.
	m := NewInMemoryMetrics()
	m.HandlerDuration("s", "c", 29)
	salida := m.Render()
	for _, want := range []string{`le="10"} 0`, `le="30"} 1`} {
		if !strings.Contains(salida, want) {
			t.Errorf("falta %q en:\n%s", want, salida)
		}
	}
}

func TestGaugeSinEtiquetasNoDejaLlavesVacias(t *testing.T) {
	m := NewInMemoryMetrics()
	m.ConnectionState(StateConnected)
	if !strings.Contains(m.Render(), "flux_connection_state 1\n") {
		t.Errorf("se esperaba `flux_connection_state 1`, se obtuvo:\n%s", m.Render())
	}
}

func TestLineasTienenFormaValidaDePrometheus(t *testing.T) {
	for _, linea := range strings.Split(strings.TrimRight(conTodasLasMetricas().Render(), "\n"), "\n") {
		if linea == "" || strings.HasPrefix(linea, "#") {
			continue
		}
		if !lineaPrometheus.MatchString(linea) {
			t.Errorf("línea no válida para Prometheus: %q", linea)
		}
	}
}

func TestEscapaLasComillasDeLosValoresDeEtiqueta(t *testing.T) {
	// Un `code` con comillas rompería el formato de exposición y Prometheus descartaría
	// el scrape ENTERO, no solo esa línea.
	m := NewInMemoryMetrics()
	m.EventDLQ("s", "c", DLQReasonPermanent, `con "comillas" y \ barra`)
	m.EventDLQ("s", "c", DLQReasonPoison, "dos\nlíneas")

	lineas := 0
	for _, linea := range strings.Split(strings.TrimRight(m.Render(), "\n"), "\n") {
		if strings.HasPrefix(linea, "#") {
			continue
		}
		lineas++
		if !lineaPrometheus.MatchString(linea) {
			t.Errorf("el escapado no produjo una línea válida: %q", linea)
		}
	}
	// Dos series, dos líneas: el salto de línea del `code` no puede partir la exposición.
	if lineas != 2 {
		t.Errorf("%d líneas, se esperaban 2", lineas)
	}
}

func TestRecolectorVacioNoExponeNada(t *testing.T) {
	// Ni cabeceras `# TYPE` de métricas que nunca se han observado: un scrape con
	// familias vacías confunde más de lo que informa.
	if got := NewInMemoryMetrics().Render(); got != "" {
		t.Errorf("un recolector vacío expuso:\n%s", got)
	}
}

func TestRecolectorEsSeguroDesdeVariosGoroutines(t *testing.T) {
	// El bucle de consumo registra mientras el handler de /metrics renderiza. Sin
	// candado, `go test -race` lo caza aquí y no en producción.
	m := NewInMemoryMetrics()
	var wg sync.WaitGroup
	for i := 0; i < 8; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < 50; j++ {
				m.EventPublished(subjectPrueba, PublishOK)
				m.HandlerDuration(subjectPrueba, consumerPrueba, 0.01)
				m.ConnectionState(StateReconnecting)
				_ = m.Render()
			}
		}()
	}
	wg.Wait()

	counters, _ := m.Snapshot()
	if got := counters[`flux_events_published_total{outcome="ok",subject="`+subjectPrueba+`"}`]; got != 400 {
		t.Errorf("contador = %v, se esperaban 400", got)
	}
}

func TestNoMetricsNoLanzaYNoGuardaNada(t *testing.T) {
	// Es el default: un SDK no debe imponer un backend de métricas.
	var sink MetricsSink = NoMetrics{}
	sink.EventPublished("s", PublishOK)
	sink.EventConsumed("s", "c", ConsumePoison)
	sink.HandlerDuration("s", "c", 1)
	sink.EventDLQ("s", "c", DLQReasonPoison, "X")
	sink.EventRetried("s", "c", 1)
	sink.ConsumerPending("s", "c", 0)
	sink.ConnectionState(StateDisconnected)
}

// InMemoryMetrics DEBE satisfacer la interfaz: si le faltara un método, el fallo
// aparecería al enchufarla en Connect y no aquí.
var _ MetricsSink = (*InMemoryMetrics)(nil)
var _ MetricsSink = NoMetrics{}
